apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerDate.cs

```diff
@@ -0,0 +1,279 @@
+﻿// Copyright 2022 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Globalization;
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// The SpannerDate type represents a logical calendar date, independent of time zone.
+    /// Represents dates with values ranging from January 1, 0001 A.D through December 31, 9999 A.D. in the Gregorian calendar.
+    /// </summary>
+    public readonly struct SpannerDate : IEquatable<SpannerDate>, IComparable<SpannerDate>, IComparable
+    {
+        // TODO : Once upgraded to .NET 6, have an operator for converting DateOnly to SpannerDate and vice versa.
+
+        // The latest possible date is 9999/12/31 (DateTime.MaxValue) and the earliest possible date is 0001/01/01 (DateTime.MinValue).
+        // The days between maximum and minimum date is MaxNumberOfDays.
+        private const int MaxNumberOfDays = 3652058;
+
+        /// <summary>
+        /// Gets the latest possible date that can be created.
+        /// </summary>
+        /// <value>
+        /// The maximum date as SpannerDate.
+        /// </value>
+        public static SpannerDate MaxDate => new SpannerDate(DateTime.MaxValue);
+
+        /// <summary>
+        /// Gets the earliest possible date that can be created.
+        /// </summary>
+        /// <value>
+        /// The minimum date as SpannerDate.
+        /// </value>
+        public static SpannerDate MinDate => new SpannerDate(DateTime.MinValue);
+
+        // The value of SpannerDate stored as the number of days since the DateTime.MinValue day i.e., 0001/01/01.
+        private readonly int _days;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SpannerDate"/> structure to the specified DateTime.
+        /// </summary>
+        /// <param name="dateTime">The dateTime as <see cref="DateTime"/>.</param>
+        public SpannerDate(DateTime dateTime) : this(dateTime.Subtract(DateTime.MinValue).Days)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SpannerDate"/> structure to the specified year, month, and day in the Gregorian calendar.
+        /// </summary>
+        /// <param name="year">The year (1 through the number of years in calendar).</param>
```

Comment:
I'd make this 1-9999, and the below 1-12. We're only supporting the Gregorian calendar, so we don't need to be circumspect about it.

(I think "in the range 1-9999 inclusive" is probably better than using "1 through 9999" as well.)

Reply:
Done.